InterfaceLib/FakeResources.c

```diff
@@ -729,5 +839,278 @@ void	FakeUseResFile( int16_t resRefNum )
+		}
+		
+		if( name )
+		{
+			memcpy(name, refEntry->resourceName, sizeof(FakeStr255));
+		}
+		return;
+	}
+	
+	gFakeResError = resNotFound;
+}
+
+
+void FakeSetResInfo( Handle theResource, int16_t theID, FakeStr255 name )
+{
+	struct FakeReferenceListEntry* refEntry = NULL;
+
+	if( !theResource || !FakeFindResourceHandle( theResource, NULL, NULL, &refEntry) )
+	{
+		gFakeResError = resNotFound;
+		return;
+	}
+
+	if( (refEntry->resourceAttributes & resProtected) != 0 )
+	{
+		gFakeResError = resAttrErr;
+		return;
+	}
+
+	refEntry->resourceID = theID;
+	memcpy(refEntry->resourceName, name, sizeof(FakeStr255));
+
+	gFakeResError = noErr;
+}
+
+
+void FakeAddResource( Handle theData, uint32_t theType, int16_t theID, FakeStr255 name )
+{
+	struct FakeResourceMap* currMap = gCurrResourceMap;
+	struct FakeTypeListEntry* typeEntry = NULL;
+	struct FakeReferenceListEntry* resourceEntry = NULL;
+
+	// AddResource() only ensures that the handle is not a resource, but doesn't check whether the type/ID are already in use
+	if( !theData || FakeFindResourceHandleInMap( theData, &typeEntry, &resourceEntry, currMap ) )
+	{
+		gFakeResError = addResFailed;
+		return;
+	}
+
+	typeEntry = FakeFindTypeListEntry( currMap, theType );
+	if( !typeEntry )
+	{
+		currMap->numTypes++;
+		currMap->typeList = realloc(currMap->typeList, currMap->numTypes * sizeof(struct FakeTypeListEntry));
+
+		typeEntry = currMap->typeList + (currMap->numTypes - 1);
+		typeEntry->resourceType = theType;
+		typeEntry->numberOfResourcesOfType = 0;
+		typeEntry->resourceList = NULL;
+		
+		FakeRetainType(theType);
+	}
+
+	typeEntry->numberOfResourcesOfType++;
+
+	if( typeEntry->resourceList )
+	{
+		typeEntry->resourceList = realloc( typeEntry->resourceList, typeEntry->numberOfResourcesOfType * sizeof(struct FakeReferenceListEntry) );
+	}
+	else
+	{
+		typeEntry->resourceList = calloc(typeEntry->numberOfResourcesOfType, sizeof(struct FakeReferenceListEntry));
+	}
+	
+	resourceEntry = typeEntry->resourceList + ( typeEntry->numberOfResourcesOfType - 1 );
+	resourceEntry->resourceAttributes = 0;
+	resourceEntry->resourceID = theID;
+	memcpy(resourceEntry->resourceName, name, sizeof(FakeStr255));
+	resourceEntry->resourceHandle = theData;
+
+	currMap->dirty = true;
+
+	gFakeResError = noErr;
+}
+
+void FakeChangedResource( Handle theResource )
+{
+	struct FakeResourceMap* theMap = NULL;
+	struct FakeReferenceListEntry* theEntry = NULL;
+	if( !FakeFindResourceHandle( theResource, &theMap, NULL, &theEntry ) )
+	{
+		gFakeResError = resNotFound;
+		return;
+	}
+
+	if( (theEntry->resourceAttributes & resProtected) == 0 )
+	{
+		theMap->dirty = true;
+		gFakeResError = noErr;
+	}
+	else
+	{
+		gFakeResError = resAttrErr;
+	}
+}
+
+// NOTE: you must call DisposeHandle(theResource) manually to release the memory.  Normally,
+//       the Resource Manager will dispose the handle on update or file close, but this implementation
+//       does not track removed resource handles for later disposal.
```

Comment:
Does this mean that old Resource Manager code didn't have to call DisposeHandle(), but code using this library will?

Reply:
Usually, clients - ours included - would call DisposeHandle() right after RemoveResource(). That's not a requirement, obviously, but it's not clear to me that writing the extra tracking code (both in the resource code and the handle code) is worth the effort.

Reply:
Perhaps it isn't. I think this library aims to be a drop-in replacement for Resource Manager code, and it isn't without this tracking, but it could get added later by someone who needs it.

Reply:
It was nowhere close to meeting that aim beforehand, and it still isn't. Given that there haven't been any changes to this library in the last 5 years, I doubt anybody really cares at this point.

Honestly, if we needed to take it further I'd probably get rid of the whole handle mechanism and write it in C++,  but then it wouldn't be semi-compatible with the legacy API.